In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#* -------------------------------------------
# df.info()
# df.head()
# df.sort_values(by='order_week', ascending=False).head()
# df.sort_values(by=['category']).head()
# count_jocabs = df[df[['first''last'] == 'Jacob','Mendoza'].value_counts()
# count_jocabs

"""
  if a specific client needed to be found in this data using their 
  fist and last names a new full_name column will need to be create where
  the first and last (name) columns will be joined
"""
df['full_name'] = df['first']+ ' ' + df['last']
df[df['full_name'] == 'Jacob Mendoza']['full_name'].value_counts()

"""
  in cases where you wish to display a single data point without an index it may 
  be appropriate to use .to_string(index=False). This will convert the data point
  to a string which can then be used seamlessly in a print statement
"""
# df['full_name'].head(1).to_string(index=False)

'\n  in cases where you wish to display a single data point without an index it may \n  be appropriate to use .to_string(index=False). This will convert the data point\n  to a string which can then be used seamlessly in a print statement\n'

In [5]:
# What three item categories had the most entries?

"""
  in this case we need to count how many times each
  category is mentioned and return the top three most
  mentioned categories. a good method for handling this 
  is .value_count()

  @params subset, normalize, sort, ascending, dropna
  @returns the total times that a given value is
  counted in the data
  .value_count()
"""
categories = df[['category']].value_counts().head(3)
categories

category   
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?

"""
  now that the most mentioned categories have been identified we need to know which
  of their subcategories is mentioned the most. one way to do this is to use .group_by().
  with this we can aggregate the data using the subcategories as our collections and
  then we just need to count their values.

  @params by, axis, level, as_index, sort, group_keys, observed, dropna
  @returns aggregated data collections
  .groupby()
"""
sub_cats = df[['category','subcategory']].groupby(['subcategory']).value_counts()
sub_cats.head(3)

subcategory        category   
accounting         software       3756
audio visual       electronics     732
bathroom supplies  consumables    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
# Store the client ids of those top 5 clients in a list.
top_five = df[['full_name','client_id']].value_counts().head()
top_five

full_name        client_id
Jessica Reyes    33615        220
Angela Everett   66037        211
Bryan Myers      46820        209
Kendra Garrett   24741        207
Alexandra Young  38378        207
Name: count, dtype: int64

In [8]:
# How many total units (the qty column) did the client with the most entries order order?

"""
  There may be times that you only need a single value from a dataframe.
  in some cases it may be best to use .index to allow for finding the value
  using its placement index

  .index[0][0]
"""

top_client = df[['full_name', 'client_id']].value_counts().head(1)
top_client = pd.DataFrame(top_client)
name = top_client.index[0][0]
id = top_client.index[0][1]
total_units = df.loc[df['client_id'] == id, 'qty'].sum()
print(f'The client with the most entries is {name} with a total of {total_units} units bought! Good job {name}! Keep it up!')

The client with the most entries is Jessica Reyes with a total of 64313 units bought! Good job Jessica Reyes! Keep it up!


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [9]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

"""
  There may be times when it is best to create a new column and store
  the aggregated data of one of more columns
"""
df['total_units_price'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number           full_name  \
0     1096.80     762.71         7.50  105            1      Donald Harding   
1       24.95      15.09         1.49   21            0       Tiffany Myers   
2       13.52       7.86         1.68   39            6      Shannon Watson   
3       36.42      24.85         1.23   29            3        Nathan Baker   
4      195.10     108.17        46.43   20            1  Christina Schwartz   

   total_units_price  
0          115164.00  
1             523.95  
2             527.28  
3            1056.18  
4            3902.00

In [23]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

"""
  We need to know how much it is going to cost to ship our units and we need to store that
  shipping price. Using an if-else statement should make it possible to add the correct value to the new shipping_price column
"""
if df['unit_weight'].all() >= 50:
  df['shipping_price'] = df['unit_weight'] * 10
else:
  df['shipping_price'] = df['unit_weight'] * 7


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number           full_name  \
0     762.71         7.50  105            1      Donald Harding   
1      15.09         1.49   21            0       Tiffany Myers   
2       7.86         1.68   39            6      Shannon Watson   
3      24.85         1.23   29            3        Nathan Baker   
4     108.17        46.43   20            1  Christina Schwartz   

  total_units_price  shipping_price  
0         115164.00           52.50  
1            523.95           10.43  
2            527.28           11.76  
3           1056.18            8.61  
4           3902.00          325.01  

[5 rows x 21 columns]

In [47]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%

"""
  In cases where you are working with floats it may be necessary to reformat pandas.
  To do this you can use pd.options.display.float_format = '{:.2f}'.format
"""
df['total_price'] = df['unit_price'] + df['shipping_price'] * .0925
pd.options.display.float_format = '{:.2f}'.format

In [ ]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).



In [ ]:
# Create a column for the profit of each line using line cost and line price



## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [ ]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [ ]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
